webapp/security/views.py

```diff
@@ -333,5 +342,185 @@ def cve(cve_id):
+    cve.discovered_by = data.get("discovered_by")
+    cve.approved_by = data.get("approved_by")
+    cve.description = data.get("description")
+    cve.ubuntu_description = data.get("ubuntu_description")
+    cve.notes = data.get("notes")
+
+    if "references" in data and len(data["references"]) > 0:
+        cve.references.clear()
+        for uri in data["references"]:
+            reference = (
+                db_session.query(CVEReference)
+                .filter(CVEReference.uri == uri)
+                .first()
+            )
+            if not reference:
+                reference = CVEReference(uri=uri)
+            cve.references.append(reference)
+
+    if "bugs" in data and len(data["bugs"]) > 0:
+        cve.bugs.clear()
+        for b in data["bugs"]:
+            bug = db_session.query(Bug).filter(Bug.uri == uri).first()
+            if not bug:
+                bug = Bug(uri=b)
+            cve.bugs.append(bug)
+
+    # Packages
+    # Check if there are packages before mapping
+    if "packages" in data and len(data["packages"]) > 0:
+        cve.packages.clear()
+        for package_data in data["packages"]:
+            package = (
+                db_session.query(Package)
+                .filter(Package.name == package_data["name"])
+                .first()
+            )
+            package.releases.clear()
+            if not package:
+                package = Package(
+                    name=package_data["name"],
+                    source=package_data["source"],
+                    ubuntu=package_data["ubuntu"],
+                    debian=package_data["debian"],
+                    releases=[],
+                )
+            for release_data in package_data["releases"]:
+                package_release_status = (
+                    db_session.query(CVEReference)
+                    .filter(CVEReference.uri == uri)
+                    .first()
+                )
+                if not package_release_status:
+                    package_release_status = PackageReleaseStatus(
+                        name=release_data["name"],
+                        status=release_data["status"],
+                        status_description=release_data["status_description"],
+                    )
+
+                package.releases.append(package_release_status)
+            cve.packages.append(package)
+
+    try:
+        db_session.add(cve)
+        db_session.commit()
+    except exc.SQLAlchemyError as error:
+        return (
+            flask.jsonify({"message": error}),
+            400,
+        )
+
+    return flask.jsonify({"message": "CVE updated succesfully"}), 200
+
+
+def delete_cve(cve_id):
+    """
+    Delete a CVE from db
+    @params string: query string with the CVE id
+    """
+    cve_query = db_session.query(CVE)
+    cve = cve_query.filter(CVE.id == cve_id).first()
+
+    try:
+        db_session.delete(cve)
+        db_session.commit()
+    except exc.IntegrityError as error:
```

Comment:
Could you explain what's going on here? When does an integrity error happen and why is it a special case?

Reply:
Well, the only way I could catch the duplicate key error was through this integrity error, and this is still a wrapper of another SQL error.